python/lsst/ts/ofc/controllers/zernike_controller.py

```diff
@@ -0,0 +1,356 @@
+    ) -> None:
+        if log is None:
+            self.log = logging.getLogger(type(self).__name__)
+        else:
+            self.log = log.getChild(type(self).__name__)
+
+        # Set OFC data
+        self.ofc_data = ofc_data
+
+        # Set gains and setpoint
+        self._kp = self.ofc_data.zk_controller["kp"]
+        self._ki = self.ofc_data.zk_controller["ki"]
+        self._kd = self.ofc_data.zk_controller["kd"]
+        self._derivative_filter_coeff = self.ofc_data.zk_controller[
+            "derivative_filter_coeff"
+        ]
+        self.setpoint = np.array(self.ofc_data.zk_controller["setpoint"])
+
+        # Set initial state
+        num_zk = self.ofc_data.znmax - self.ofc_data.znmin + 1
+        self.zk_state0 = np.zeros((num_sensors, num_zk))
+        self.zk_state = self.zk_state0.copy()
+
+        # Initialize previous error and integral
+        self.previous_error = self.setpoint - self.zk_state0
+        self.integral = self.previous_error
+        self.filtered_derivative = np.zeros((num_sensors, num_zk))
+
+    @property
+    def kp(self) -> float:
+        return self._kp
+
+    @kp.setter
+    def kp(self, value: float) -> None:
+        self._kp = value
+
+    @property
+    def ki(self) -> float:
+        return self._ki
+
+    @ki.setter
+    def ki(self, value: float) -> None:
+        self._ki = value
+
+    @property
+    def kd(self) -> float:
+        return self._kd
+
+    @kd.setter
+    def kd(self, value: float) -> None:
+        self._kd = value
+
+    @property
+    def derivative_filter_coeff(self) -> float:
+        return self._derivative_filter_coeff
+
+    @derivative_filter_coeff.setter
+    def derivative_filter_coeff(self, value: float) -> None:
+        self._derivative_filter_coeff = value
+
+    def reset_zk_state(self) -> None:
+        """Initialize the state to the state 0 in the basis of degree of
+        freedom (DOF).
+        """
+        self.zk_state = self.zk_state0.copy()
+
+    def aggregate_state(
+        self, zk: np.ndarray[float] | list, zn_idx: np.ndarray[int] | list[int]
+    ) -> None:
+        """Aggregate the calculated degree of freedom (zk) in the state.
+
+        Parameters
+        ----------
+        zk : `numpy.ndarray` or `list`
+            Calculated zks.
+        zn_idx : `numpy.ndarray` or `list[int]`
+            Index array of degree of freedom.
+        """
+        self.zk_state += zk
+
+    @property
+    def aggregated_state(self) -> np.ndarray[float]:
+        """Returns the aggregated state.
+
+        Returns
+        -------
+        `np.ndarray[float]`
+            Aggregated state.
+        """
+        return self.zk_state
+
+    def set_aggregated_state(self, value: np.ndarray[float]) -> None:
+        """Set the aggregated state.
+
+        Parameters
+        ----------
+        value : `np.ndarray[float]`
+            Aggregated state.
+        """
+        self.zk_state += value
```

Comment:
Method is called `set_...` but it is doing the same as `aggregate_...`. Should this be `self.zk_state = value` instead?